### PR TITLE
unittest -> pytest

### PR DESCRIPTION
**Pull request by cjermain**
This PR migrates the testing from `unittest` to `pytest`. This allows for the use of parametrized tests and fixtures, leading to more robust test coverage and easier test-code maintenance. This was motivated by my recent PRs on Bidirectional and Masking features, where I saw that many of the tests rely on nested for-loops, which are hard to filter (using `-k`) and maintain as they expand.

Goals:
- [ ] Replace nested for-loops with parametrized tests
- [ ] Replace test running functions with fixtures
- [ ] Improve access to `actual` predictions and better integrate into pytest/pdb
- [ ] Expand test coverage with additional parametrized tests

This PR is still a work in progress, and will be updated. I'm interested in any suggestions or comments.

### CODE REVIEW

**Comment by wenbingl** [comment]
Glad to the effort, does pytest support the Test class as well?

**Comment by cjermain** [comment]
Thanks. Pytest uses functions instead of test classes. [Fixtures in pytest](https://docs.pytest.org/en/latest/fixture.html) allow for setup/teardown of test resources, and can be composed to handle more complex situations. I am most interested in using [parametrized tests](https://docs.pytest.org/en/latest/parametrize.html) so that we can loop over a large number of cases to get better test coverage.

File: tests/test_layers.py [hunk 1]
```diff
@@ -1308,788 +1397,807 @@ def _batch_norm_helper(self, data, gamma, beta, scale, center, axis):
...
-                expected = model.predict(x)
-                self.assertTrue(run_onnx_runtime(onnx_model.graph.name, onnx_model, x, expected, self.model_files))
+        input2 = Input(shape=(None, 5))
+        states = Bidirectional(rnn_class(2, return_state=True))(input1)[1:]
+        out = Bidirectional(rnn_class(2, return_sequences=True))(input2, initial_state=states)
+        model = Model([input1, input2], out)
+        inputs = [x, x]
 
-    def test_rnn_state_passing(self):
-        K.clear_session()
```
Comment (cjermain) [inline tests/test_layers.py:1618]: This is being run for all tests now in the `runner` fixture. It helps avoid difficult to reproduce errors that occur due to resource leaking in the tests.

**Comment by cjermain** [comment]
This PR is ready for review. All of the main tests are converted to using pytest. There are a number of additional goals above, but given the diff already, I think they are better left for other PRs. This includes migrating the nightly tests, which still use unittest.

I would highly recommend using the "Hide whitespace changes" to review this PR, since the majority of the changes are indentation.
Screenshot_2020-04-12  WIP  unittest - pytest by cjermain · Pull Request #425 · onnx keras-onnx

@wenbingl & @jiafatom, can you review this PR?

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+###############################################################################
```
Comment (wenbingl) [inline tests/conftest.py:1]: what's the meaning of 'conftest'?

File: tests/test_cgan.py [hunk 3]
```diff
@@ -120,29 +119,15 @@ def build_discriminator(self):
...
-        onnx_model = keras2onnx.convert_keras(keras_model, keras_model.name)
-        self.assertTrue(run_onnx_runtime(onnx_model.graph.name, onnx_model,
-                                         {keras_model.input_names[0]: x, keras_model.input_names[1]: y}, expected,
-                                         self.model_files))
-
-
-if __name__ == "__main__":
-    unittest.main()
+@pytest.mark.skipif(keras2onnx.proto.tfcompat.is_tf2 and is_tf_keras, reason="Tensorflow 1.x only tests.")
+@pytest.mark.skipif(is_tf_keras and StrictVersion(tf.__version__) < StrictVersion("1.14.0"),
```
Comment (wenbingl) [inline tests/test_cgan.py:123]: StrictVersion(tensorflow.__version__.split('-')[0]) >= StrictVersi...
otherwise, the whole test cases will failed on tensorflow nightly build.

File: tests/test_misc.py [hunk 4]
```diff
@@ -10,47 +15,41 @@
...
+    converted_model_5 = keras2onnx.convert_keras(model, target_opset=5)
 
-        opset_comparison = converted_model_7.opset_import[0].version > converted_model_5.opset_import[0].version
+    assert converted_model_7 is not None
+    assert converted_model_5 is not None
 
-        self.assertTrue(opset_comparison)
-
-
-if __name__ == '__main__':
```
Comment (wenbingl) [inline tests/test_misc.py:55]: we sometimes run this single test file with python cmd line instead of pytest, which helps on debugging. Is there alternative way in the pytest as well?

File: tests/conftest.py (hunk 2 above)
Comment (cjermain) [reply tests/conftest.py:1]: `conftest.py` is the way that `pytest` shares fixtures across the entire test suite.
https://docs.pytest.org/en/latest/fixture.html#conftest-py-sharing-fixture-functions

File: tests/test_misc.py (hunk 4 above)
Comment (cjermain) [reply tests/test_misc.py:55]: You can select a specific file by running `pytest` against that file.
```
pytest tests/test_misc.py -vv
```
You can also filter tests for specific strings of interest (e.g. LSTM):
```
pytest tests -k LSTM -vv
```

File: tests/test_cgan.py (hunk 3 above)
Comment (cjermain) [reply tests/test_cgan.py:123]: I'm confused by what you mean here. The current conditional is the same as that in line 133, so the previous behavior is reproduced. Was the original incorrect?

**Comment by cjermain** [comment]
> Can you run code format to format the code before merge? like 'Reformat code' in pycharm.

I use `vim` and don't have a formatter built in to my editor. There are a number of CLI formatting tools out there, for example [black](https://github.com/psf/black), [yapf](https://github.com/google/yapf), and [autopep8](https://github.com/hhatto/autopep8). Can you be more specific about the formatting you are looking for?

File: tests/test_cgan.py (hunk 3 above)
Comment (wenbingl) [reply tests/test_cgan.py:123]: The previous code is also ‘imprecise’, but the issue was hidden because this condition wasn't tested due to the first is_tf2 test is already wrong. And it looks the pytest.mark will test both. If you install a tensorflow nightly build. you will see the failure.

**Comment by wenbingl** [comment]
> > Can you run code format to format the code before merge? like 'Reformat code' in pycharm.
> 
> I use `vim` and don't have a formatter built in to my editor. There are a number of CLI formatting tools out there, for example [black](https://github.com/psf/black), [yapf](https://github.com/google/yapf), and [autopep8](https://github.com/hhatto/autopep8). Can you be more specific about the formatting you are looking for?

I saw there were few PEP8 suggestions in my editor tool. But it's not mandatory now. Feel free to ignore it if it is not straightforward in your development.

File: tests/test_cgan.py (hunk 3 above)
Comment (cjermain) [reply tests/test_cgan.py]: Ok, that makes sense. I've made the change.

**Comment by wenbingl** [comment]
Thanks a lot for your contribution!
